docs: 3d mesh finetuning tutorial (#345)
jina-ai/finetuner: docs/get-started/3d-mesh/index.md

# Finetuning PointConv on ModelNet40 Dataset

In this tutorial we will finetune PointConv on [ModelNet40 dataset](https://arxiv.org/pdf/1406.5670v3.pdf).
The ModelNet40 dataset contains synthetic object point clouds. As the most widely used benchmark for point cloud analysis, ModelNet40 is popular because of its various categories, clean shapes, well-constructed dataset.

> 

Reviewer: I think original is better than the suggestion

```{tip}
ModelNet: a consistent, large-scale dataset of 3D objects annotated with fine-grained, instance-level, and hierarchical 3D part information.
The corresponding point cloud data points are uniformly sampled from the mesh surfaces, and then further preprocessed by moving to the origin and scaling into a unit sphere. This dataset is used to establish three benchmarking tasks for evaluating 3D part recognition: fine-grained semantic segmentation and classification. In our case we'll use it to finetune PointConv which was pretrained on [ShapeNet dataset](https://arxiv.org/abs/1512.03012).
```

You can download ModelNet dataset on the official [Princeton website](https://modelnet.cs.princeton.edu/).

```shell
wget http://modelnet.cs.princeton.edu/ModelNet40.zip

```

## Data preparation

After downloading ModelNet data we need to unzip it.

```shell
pip install finetuner
pip install trimesh

unzip ModelNet40.zip
```

Before we go further let's look at some examples from our data:

```{figure} airplane.gif
:align: center
:width: 70%
```

```{figure} lap.gif
:align: center
:width: 70%
```

> Now let's go to loading our data. We'll use `DocumentArray` from the library `docarray`. We'll go through all the files in our directory.

Reviewer 2: We'll go through all the files in our directory and we'll recursively read all files that end with `.off`

all the files in our directory and we'll recursively read all files that end with .off.

> We split the files between two `DocumentArray`s depending on their uri to train and test.

Reviewer 3: ```suggestion
We split the files into two `DocumentArray`s depending on their uri to train and test.
```

Finally we save the resulting `DocumentArray`s into binary files for later use.

```python
import glob
import os
from typing import Optional

import trimesh
from docarray import Document, DocumentArray

train = DocumentArray()
test = DocumentArray()
data_path = 'ModelNet40'
all_docs = DocumentArray.from_files(os.path.join(data_path,'/**/**/*.off'))
for doc in all_docs:

>     if 'test' in doc.uri :

Reviewer 3: ```suggestion
    if 'test' in doc.uri:
```

        test.append(doc)
    else:
        train.append(doc)

train.save_binary('train_data_modelnet.bin')
test.save_binary('test_data_modelnet.bin')

```

## Model Training

> Now that we have our data ready, we need a model that creates embeddings we will later use those embeddings for searching similar matches for different queries. A lot of work and research has been done in the field of 3D data embeddings there are some powerful models like [PointConv](https://arxiv.org/abs/1811.07246) which have been trained on [ShapeNet dataset](https://arxiv.org/pdf/1512.03012.pdf)

Reviewer 2: we need a model that creates embeddings so that we can later use these embeddings

> Now that we have our data ready, we need a model that creates embeddings we will later use those embeddings for searching similar matches for different queries. A lot of work and research has been done in the field of 3D data embeddings there are some powerful models like [PointConv](https://arxiv.org/abs/1811.07246) which have been trained on [ShapeNet dataset](https://arxiv.org/pdf/1512.03012.pdf)

Reviewer 2: A lot of work and research has been done in the field of 3D data embeddings **and** there are some powerful models

> In order to not reivent the wheel we will use this model. We will also use Jina's 3D Mesh Encoder which wraps these two models in an executor.

Reviewer 2: reinvent

> This executor receives Documents containing point sets data in its blob attribute, with shape (N, 3) and encodes it to embeddings of shape (D,). Now, the following pretrained models are ready to be used to create embeddings:

Reviewer 2: and encodes them

> - PointConv-Shapenet-d512: A PointConv model resulted in 512 dimension of embeddings, which is finetuned based on ShapeNet dataset.

Reviewer 2: finetuned on ShapeNet dataset

> - PointConv-Shapenet-d512: A PointConv model resulted in 512 dimension of embeddings, which is finetuned on ShapeNet dataset.
> - PointConv-Shapenet-d1024: A PointConv model resulted in 1024 dimension of embeddings, which is finetuned on ShapeNet dataset.

Reviewer 3: ```suggestion
- PointConv-Shapenet-d512: A PointConv model, finetuned on ShapeNet dataset, resulting in 512 dimension of embeddings
- PointConv-Shapenet-d1024: A PointConv model,  finetuned on ShapeNet dataset, resulting in 1024 dimension of embeddings,.
```

> - PointConv-Shapenet-d1024: A PointConv model resulted in 1024 dimension of embeddings, which is finetuned based on ShapeNet dataset.

Reviewer 2: finetuned on ShapeNet dataset

```python
def random_sample(pc, num):
    permutation = np.arange(len(pc))
    np.random.shuffle(permutation)
    pc = np.array(pc).astype('float32')
    pc = pc[permutation[:num]]
    return pc

def preprocess(doc: 'Document', num_points: int = 1024, data_aug: bool = True):
    doc.load_uri_to_point_cloud_tensor(2048)
    doc.tags['finetuner_label'] = doc.uri.split('/')[2]
    points = random_sample(doc.tensor, num_points)

    points = points - np.expand_dims(np.mean(points, axis=0), 0)  # center
    dist = np.max(np.sqrt(np.sum(points ** 2, axis=1)), 0)
    points = points / dist  # scale

    if data_aug:
        theta = np.random.uniform(0, np.pi * 2)
        rotation_matrix = np.array(
            [[np.cos(theta), -np.sin(theta)], [np.sin(theta), np.cos(theta)]]
        )
        points[:, [0, 2]] = points[:, [0, 2]].dot(rotation_matrix)  # random rotation
        points += np.random.normal(0, 0.02, size=points.shape)  # random jitter
    return points
```

In the following code snippet we create MeshData model which encapsulates a `PointConv` model with 512 dimensions, we then load

> training and evaluation data from the binary fs we saved before. We create an optimizer and a learning rate scheduler. In this case with use

Reviewer 2: fs = files?

an Adam optimizer and MultiStepLR scheduler but you can change those depending on your data and preferences.

```python
model = MeshDataModel(model_name='pointconv', embed_dim=512) # create pointconv model with 512 dimensions

train = DocumentArray.load_binary('train_data_modelnet.bin') # load train dataset

> eval = DocumentArray.load_binary('test_data_modelnet.bin') # load eval dataset

Reviewer 2: eval is reserved in Python, use val or dev

def configure_optimizer(model):
    from torch.optim import Adam
    from torch.optim.lr_scheduler import MultiStepLR
    # create Adam optimizer with MultistepLR scheduler
    optimizer = Adam(model.parameters(), lr=5e-4)
    scheduler = MultiStepLR(optimizer, milestones=[30, 60], gamma=0.5)

    return optimizer, scheduler
```

> We leverage finetuner's mining strategy to imprive and finetuning the embeddings. By this we mean that the finetuner will use

Reviewer 2: to improve and finetune

> We leverage finetuner's mining strategy to improve and finetune the embeddings. By this we mean that the finetuner will use

Reviewer 3: Since we are using `TripletEasyHardMiner`, can you explain why it is called triplet and how to get the anchor-positive-negative triplets?

Author: I don't want to give a small and insufficient answer so I added a phrase if they want to learn more I redirect them to that section in finetuner docs

the mentioned label coming with every `Document` which we specified during preprocessing and will sample it as a positive example
and then sample examples with different label as negatives.

The triplet loss's goal is to pull `Documents` with same class together and those with different classes away from each other thus improving
the final representation.

```python

> tuned_model = finetuner.fit(

Reviewer: Can you add WandB callback for logging here?

**mNDCG@k**: We'll calculate NDCG at 1,5 and 10, then we'll calculate the mean of those for all documents in our test data  because we care about the order of the retrieved 3D objects.

````{dropdown} Complete source code

```python
from finetuner.tuner.evaluation import Evaluator

train = DocumentArray.load_binary('../train_data_modelnet.bin') # load train dataset
eval = DocumentArray.load_binary('../test_data_modelnet.bin') # load test dataset

train.apply(preprocess)
eval.apply(preprocess)

tuned_model = MeshDataModel(model_name='pointconv', embed_dim=512) # create pointconv model with 512 dimensions
pretrained_model = MeshDataModel(model_name='pointconv', embed_dim=512)

tuned_model.load_state_dict(torch.load('checkpoints/finetuned-pointconv-d512.pth')) # load finetuned weights

tuned_model.eval()
pretrained_model.eval()

evaluator_finetuned = Evaluator(eval, train, tuned_model)
print(evaluator_finetuned.evaluate())

evaluator_pretrained = Evaluator(eval, train, pretrained_model)
print(evaluator_pretrained.evaluate())
```
````

The difference is shown in the tables below:

| mAP@k  | pre-trained | fine-tuned |
|--------|-------------|------------|

> | mAP@1  | 0.147       | 0.719      |

Reviewer 2: how do you extract metrics at 1, 5, 10? The evaluator can't do that yet

Author: Using evaluator I get loop disconnect from AWS don't know why, that was from a custom script

| mAP@5  | 0.113       | 0.697      |
| mAP@10 | 0.100       | 0.686      |

| mNDCG@k  | pre-trained | fine-tuned |
|--------|-------------|--------------|
| mNDCG@1  | 0.563       | 0.927      |
| mNDCG@5  | 0.617       | 0.931      |
| mNDCG@10 | 0.647       | 0.935      |

Now let's do some queries ourselves and check the visualizations

````{dropdown} Complete source code

```python
train_da = DocumentArray.load_binary('../train_data_modelnet.bin') # load train dataset
eval_da = DocumentArray.load_binary('../test_data_modelnet.bin') # load eval dataset

train_da.apply(preprocess)
eval_da.apply(preprocess)

tuned_model = MeshDataModel(model_name='pointconv', embed_dim=512) # create pointconv model with 512 dimensions
tuned_model.load_state_dict(torch.load('checkpoints/finetuned-pointconv-d512.pth'))
tuned_model.eval()

train_da.embed(tuned_model, batch_size=128, device='cuda')
eval_da.embed(tuned_model, batch_size=128, device='cuda')

eval_da.match(train_da, limit=10)
```
````

> Query:

Reviewer: Can you check if it is possible to create a table here, so that query, pretrained and finetuned results can be shown side to side? This would improve the presentation I think

Author: I tried different solutions including html, I can't seem to get it done

Reviewer 3: I don't know if you have tried this already but it may help: https://stackoverflow.com/a/24320279
I was able to show side by side in my PyCharm 
![image](https://user-images.githubusercontent.com/6537525/151143901-da335e8c-66cf-4c8b-9fe1-8800b7a00b73.png)

Author: are they still gifs ? I tried it and it didn't work, can you perhaps share code

Reviewer 3: Yes they are still rendered as gif. Code is taken from the above link mentioned, I am attaching the snapshot.

![Screenshot 2022-01-26 at 13 19 56](https://user-images.githubusercontent.com/6537525/151161932-aea24f64-b1fc-46aa-85bc-302ab363d15b.png)

https://user-images.githubusercontent.com/6537525/151161605-50c54949-5007-4f31-85e6-36ee510aadb5.mov

Author: this is what I'm getting locally
![Screenshot from 2022-01-26 15-05-22](https://user-images.githubusercontent.com/40893766/151177530-6e0630d7-cc1c-40d9-a920-6a2349416a3b.png)
 
![Screenshot from 2022-01-26 15-06-43](https://user-images.githubusercontent.com/40893766/151177598-3cb03622-96ce-4b60-a336-d8da5570ced6.png)

```{figure} radio_query.gif
:align: center
:width: 70%
```

Pretrained PointConv:

```{figure} radio_oob.gif
:align: center
:width: 70%
```

Finetuned PointConv:

```{figure} radio_finetuned.gif
:align: center
:width: 70%
```

Query:

```{figure} dresser_query.gif
:align: center
:width: 70%
```

Pretrained PointConv:

```{figure} dresser_oob.gif
:align: center
:width: 70%
```

Finetuned PointConv:

```{figure} dresser_finetuned.gif
:align: center
:width: 70%
```

> We can clearly see that after finetuning PointConv has enhanced embeddings, and that's how you finetune a 3D model.

Reviewer 2: `and that's how you finetune a 3D model.` insert thug life meme

the final representation.

```python
tuned_model = finetuner.fit(
    model,
    train_da,
    eval_data=eval_da,
    preprocess_fn=partial(preprocess, num_points=2048, data_aug=True),
    epochs=epochs,
    batch_size=batch_size,
    loss=TripletLoss(
        miner=TripletEasyHardMiner(pos_strategy='easy', neg_strategy='semihard')
    ),
    configure_optimizer=configure_optimizer,
    learning_rate=5e-4,
    device='cuda' if use_gpu else 'cpu',
)
# saving the finetuner model
torch.save(
    tuned_model.state_dict(),
    str(checkpoint_dir / f'finetuned-{model_name}-d{embed_dim}.pth'),
)
```

## All in one script

We provide all the code above in one script so it's super easy to use. This script is inside the repository that we clone earlier and here's how you can
use it:

```shell
python executor-3d-encoder/finetune.py --model_name pointconv \
                   --train_dataset ../train_data_modelnet.bin \
                   --eval_dataset ../test_data_modelnet.bin \
                   --batch_size 64 \
                   --epochs 50 \
                   --use-gpu

> ```

Reviewer 4: again, i don't think it's needed, brings limited value to the tutorial

````{dropdown} Complete source code

```python
train_da = DocumentArray.load_binary('../train_data_modelnet.bin') # load train dataset
eval_da = DocumentArray.load_binary('../test_data_modelnet.bin') # load test dataset

train_da.apply(preprocess)
eval_da.apply(preprocess)

tuned_model = MeshDataModel(model_name='pointconv', embed_dim=512) # create pointconv model with 512 dimensions
tuned_model.load_state_dict(torch.load('checkpoints/finetuned-pointconv-d512.pth')) # load finetuned weights
tuned_model.eval()

train_da.embed(tuned_model, batch_size=128, device='cuda') # create embeddings
eval_da.embed(tuned_model, batch_size=128, device='cuda')

eval_da.match(train_da, limit=10)

def mean_average_precision_at_k(da, topk=1):
    hit = []
    avg_pr = []
    for d in da:
        for m in d.matches[:topk]:
            if d.uri.split('/')[-1].split('_')[0] == m.uri.split('/')[-1].split('_')[0]:
                hit.append(1)
            else:
                hit.append(0)
        avg_pr.append(average_precision(hit))

>     return np.mean(avg_pr)

Reviewer 4: use built-in evaluator

```
````

The difference is shown in the tables below:

| mAP@k  | pre-trained | fine-tuned |
|--------|-------------|------------|
| mAP@1  | 0.147       | 0.719      |
| mAP@5  | 0.113       | 0.697      |
| mAP@10 | 0.100       | 0.686      |

| mNDCG@k  | pre-trained | fine-tuned |
|--------|-------------|--------------|
| mNDCG@1  | 0.563       | 0.927      |
| mNDCG@5  | 0.617       | 0.931      |
| mNDCG@10 | 0.647       | 0.935      |

Now let's do some queries ourselves and check the visualizations

````{dropdown} Complete source code

```python
train_da = DocumentArray.load_binary('../train_data_modelnet.bin') # load train dataset
eval_da = DocumentArray.load_binary('../test_data_modelnet.bin') # load eval dataset

train_da.apply(preprocess)
eval_da.apply(preprocess)

tuned_model = MeshDataModel(model_name='pointconv', embed_dim=512) # create pointconv model with 512 dimensions
tuned_model.load_state_dict(torch.load('checkpoints/finetuned-pointconv-d512.pth'))
tuned_model.eval()

train_da.embed(tuned_model, batch_size=128, device='cuda')
eval_da.embed(tuned_model, batch_size=128, device='cuda')

eval_da.match(train_da, limit=10)

> visualize(eval_da.matches[0].uri)

Reviewer 4: this `visualize` function does not exist anymore I think.

The original ModelNet40 consists of 12,311 CAD-generated meshes in 40 categories (such as airplane, car, plant, lamp), of which 9,843 are used for training while the rest 2,468 are reserved for testing.

```{tip}
ModelNet: a consistent, large-scale dataset of 3D objects annotated with fine-grained, instance-level, and hierarchical 3D part information.
The corresponding point cloud data points are uniformly sampled from the mesh surfaces, and then further preprocessed by moving to the origin and scaling into a unit sphere. This dataset is used to establish three benchmarking tasks for evaluating 3D part recognition: fine-grained semantic segmentation and classification. In our case we'll use it to finetune PointConv which was pretrained on [ShapeNet dataset](https://arxiv.org/abs/1512.03012).
```

You can download ModelNet dataset on the official [Princeton website](https://modelnet.cs.princeton.edu/).

```shell
wget http://modelnet.cs.princeton.edu/ModelNet40.zip

```

## Data preparation

After downloading ModelNet data we need to unzip it.

```shell
pip install docarray
pip install finetuner

> pip install trimesh

Reviewer 5: `docarray` would not be needed, since it has been included in finetuner

                    method='animate',
                    args=[None, dict(frame=dict(duration=50, redraw=True),
                        transition=dict(duration=0),
                        fromcurrent=True,
                        mode='immediate'
                        )]
                    )
                ])]
        ),
        frames=frames
    )

    return fig

def pcshow(xs,ys,zs):
    data=[go.Scatter3d(x=xs, y=ys, z=zs,
                                   mode='markers')]
    fig = visualize_rotate(data)
    fig.update_traces(marker=dict(size=2,
                      line=dict(width=2,
                      color='DarkSlateGrey')),
                      selector=dict(mode='markers'))
    fig.show()

def visualize(uri):
    with open(uri, 'r') as f:
        verts, faces = read_off(f)
        
    i,j,k = np.array(faces).T
    x,y,z = np.array(verts).T
    visualize_rotate([go.Mesh3d(x=x, y=y, z=z, color='gray', opacity=0.5, i=i,j=j,k=k)]).show()
```
````

```{figure} airplane.gif
:align: center
:width: 70%
```

```{figure} lap.gif
:align: center
:width: 70%
```

```{figure} vase.gif
:align: center
:width: 70%
```

```{figure} person.gif
:align: center
:width: 70%
```

> Now let's go to loading our data. We'll use `DocumentArray` from the library `docarray`. We'll go through all the files in our directory.

Reviewer 4: ```suggestion
Now let's go to load our data. We'll use `DocumentArray` from the library `docarray`. We'll go through all the files in our directory.
```

Along with loading point clouds, we also assign the `finetuner_label` as a tag. This label will be later used during finetuning. Finally we save
the resulting `DocumentArray` into a binary file for later use. This way we don't have to convert our data to point clouds again.

```python
import glob
import os
from typing import Optional

import trimesh
from docarray import Document, DocumentArray

train_docs = DocumentArray()
test_docs = DocumentArray()
data_path = 'ModelNet40'
for dir in os.listdir(data_path):
    dir_path = os.path.join(data_path,dir)
    for file in glob.iglob(os.path.join(dir_path,'train/*.off'),recursive=True):
            doc= Document(uri=file)
            doc.load_uri_to_point_cloud_tensor(2048)
            doc.tags['finetuner_label'] = dir
            train_docs.append(doc)
    for file in glob.iglob(os.path.join(dir_path,'test/*.off'),recursive=True):
            doc= Document(uri=file)
            doc.load_uri_to_point_cloud_tensor(2048)
            doc.tags['finetuner_label'] = dir
            test_docs.append(doc)

train_docs.save_binary('train_data_modelnet.bin')

> test_docs.save_binary('test_data_modelnet.bin')

Reviewer 4: `dir` and `file` are python reserved keywords, besides, can we do:
```python
train = DocumentArray.from_files(..)
test = DocumentArray.from_files(..)

def preprocess(doc):
    doc.load_uri_to_point_cloud_tensor(..)
    doc.tags[..] = doc.uri
    return doc

train.apply(..)
test.apply(..)
...

```
